java/rocksjni/portal.h

```diff
@@ -4474,25 +4458,25 @@ class JniUtil {
       return jbyte_strings;
     }
 
-    /**
-     * Copies bytes from a rocksdb::Slice to a jByteArray
-     *
-     * @param env A pointer to the java environment
-     * @param bytes The bytes to copy
-     *
-     * @return the Java byte[] or nullptr if an exception occurs
-     */
-    static jbyteArray copyBytes(JNIEnv* env, const Slice& bytes) {
-      const jsize jlen = static_cast<jsize>(bytes.size());
-
+    static jbyteArray createJavaByteArrayWithSizeCheck(JNIEnv* env, const char* bytes, const size_t size) {
+      // Limitation for array size is vm specific
+      // Current HotSpot VM limitation for array size is Integer.MAX_VALUE - 5 (2^31 - 1 - 5)
+      // Integer.MAX_VALUE - 8 should be safe enough
+      static const size_t MAX_JARRAY_SIZE = ((static_cast<size_t>(1)) << 31) - 9;
+      if(size > MAX_JARRAY_SIZE) {
+        rocksdb::RocksDBExceptionJni::ThrowNew(env, "Requested array size exceeds VM limit");
```

Comment:
The though occurs to me that, we are now potentially throwing a `RocksDBException` for any caller of `copyBytes` which we were not previously.

So we need to look for each Java function that eventually calls this code and make sure it has an `@throws RocksDBException` on both itself and its native interface function. A quick `grep -r "copyBytes" java/rocksdbjni` shows that we will in the least need changes to `WriteBatch.Handler` functions, amongst probably others.

Reply:
As I understand, `WriteBatch.Handler` methods  are called from JNI code which is behind `WriteBatch.iterate` (marked as throws `RocksDBException`). `copyBytes` is called from there before, to prepare key/value for `WriteBatch.Handler`. `WriteBatch.Handler` methods  are user specific java code.

I also found couple of other places which can reach `copyBytes`.